pip/pip-376.md

```diff
@@ -0,0 +1,138 @@
+
+## Badly designed TopicPoliciesService interface
+
+The `TopicPoliciesService` interface is a terrible abstraction because it's never designed for 3rd party implementations.
+
+1. Methods that should not be exposed
+
+`addOwnedNamespaceBundleAsync` and `removeOwnedNamespaceBundleAsync` are only used internally in `SystemTopicBasedTopicPoliciesService`.
+
+`getTopicPoliciesBypassCacheAsync` is only used in tests. This method just creates a reader to replay the `__change_events` topic and construct the topic policies map.
+
+2. Confusing and inconsistent `getTopicPolicies` family
+
+There are two overrides of `getTopicPolicies`:
+
+```java
+TopicPolicies getTopicPolicies(TopicName topicName, boolean isGlobal) throws TopicPoliciesCacheNotInitException;
+TopicPolicies getTopicPolicies(TopicName topicName) throws TopicPoliciesCacheNotInitException;
+```
+
+The 2nd method is equivalent to `getTopicPolicies(topicName, false)`.
+
+The semantics of these two methods are very intuitive. First, they are not synchronous methods that are blocked by waiting a future. They just start an asynchronous policies initialization (creating a reader to replay the `__change_events` topic), and then try to get the policies from the cache. If the asynchronous policies initialization didn't start, just throw `TopicPoliciesCacheNotInitException`.
+
+As you can see, these two methods are hard to use. And they are also only used in tests except for the `getTopicPoliciesAsyncWithRetry` method, which uses a user-provided executor and backoff policy to call `getTopicPolicies` until `TopicPoliciesCacheNotInitException` is not thrown:
+
+```java
+    default CompletableFuture<Optional<TopicPolicies>> getTopicPoliciesAsyncWithRetry(TopicName topicName,
+              final Backoff backoff, ScheduledExecutorService scheduledExecutorService, boolean isGlobal) {
+```
+
+The `getTopicPolicies` overrides are only called in tests while `getTopicPoliciesAsyncWithRetry` is used in the core. It would be very confusing to users that want to implement their own topic policies service. They have to look deeply into the Pulsar's source code to know these details.
+
+https://github.com/apache/pulsar/pull/21231 adds two asynchronous overrides that are much more friendly to users:
+
+```java
+CompletableFuture<Optional<TopicPolicies>> getTopicPoliciesAsync(@Nonnull TopicName topicName, boolean isGlobal);
+CompletableFuture<Optional<TopicPolicies>> getTopicPoliciesAsync(@Nonnull TopicName topicName);
+```
+
+Now we have **5** asynchronous get methods. What's worse, unlike `getTopicPolicies`, `getTopicPoliciesAsync(topic)` is not equivalent to `getTopicPoliciesAsync(topic, false)`, instead,
+- `getTopicPoliciesAsync(topic)` will try getting local policies first, if absent, then try getting global policies
```

Comment:
Is it better to not expose this method? Users should only implement 

```
CompletableFuture<Optional<TopicPolicies>> getTopicPoliciesAsync(@Nonnull TopicName topicName, boolean isGlobal);
```

Let's the broker decide when to get local topic policy or global policy. At lease for now, I don't think users should customize this behavior

Reply:
> Let's the broker decide

That's why I design the `getTopicPoliciesAsync` interface like

```java
CompletableFuture<Optional<TopicPolicies>> getTopicPoliciesAsync(TopicName topicName, GetType type);
```

- If broker side just want to get policies according to the priority, it will call `getTopicPoliciesAsync(topic, LOCAL_FIRST)`.
- If broker side only want to get local policies or global policies, it will call `getTopicPoliciesAsync(topic, LOCAL_ONLY)` or `getTopicPoliciesAsync(topic, GLOBAL_ONLY)`.

See
- https://github.com/BewareMyPower/pulsar/blob/02bffa26332ee48212c50eb985aeb683825f9055/pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/AdminResource.java#L367-L371
- https://github.com/BewareMyPower/pulsar/blob/02bffa26332ee48212c50eb985aeb683825f9055/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java#L1176-L1179